v3/src/liftoff/core.test.ts

```diff
@@ -0,0 +1,75 @@
+import { Core } from "./core";
+import { str, int, obj, func } from "./ref";
+import { def } from "./pattern";
+
+const USER_SCHEMA = {
+  typeDefs: `
+    type User {
+      uid: String
+      name: String
+    }
+  `,
+  resolvers: {
+    User() {
+      return { uid: "0", name: "root" };
+    }
+  }
+};
+
+const PRODUCT_SCHEMA = {
+  typeDefs: `
+    type Product {
+      id: Integer
+      title: String
+    }
+  `,
+  resolvers: {
+    Product() {
+      return { id: 0, title: "A sock" };
+    }
+  }
+};
+
+const MAX_CONNECTIONS_DEFAULT = 1024;
+const SERVICE_NAME = "Example Service";
+
+describe("the core", () => {
+  const serviceName = str`Name of the service`();
+  const maxConnections = int`Maximum connections`(MAX_CONNECTIONS_DEFAULT);
+  const preprocess = func<(input: string) => string>`Query preprocessor`();
+  const Schema = obj<{ typeDefs: string; resolvers: any }>`GraphQL Schema`();
+
+  const core = new Core(() => {
+    def(Schema)(PRODUCT_SCHEMA);
+    def(serviceName)(SERVICE_NAME);
+    def(Schema)(USER_SCHEMA);
+  });
+
+  describe(".once", () => {
+    it("returns an array of all ref definitions", async () => {
+      expect(await core.once(Schema)).toEqual([PRODUCT_SCHEMA, USER_SCHEMA]);
+    });
+
+    it("returns the defaultValue if the ref was never defined", async () => {
+      expect(await core.once(maxConnections)).toEqual([
```

Comment:
Oh, hmm… So in this case it's because we're testing what happens when a ref isn't defined but has a default value, and `maxConnections` is the only ref that fits that criteria—we test the other refs, which express different conditions, in those tests.

The tests could certainly be more exhaustive. We're not testing each of these ref-definition conditions for each ref type, but that didn't seem super important, since the machinery is identical (the different ref creators are really just sugar).

But it isn't super clear. Would some comments where we instantiate the core be helpful?